lib/parallel.rb

```diff
@@ -97,20 +97,26 @@ def wait
 
   class JobFactory
     def initialize(source, mutex)
-      @lambda = (source.respond_to?(:call) && source) || queue_wrapper(source)
-      @source = source.to_a unless @lambda # turn Range and other Enumerable-s into an Array
+      @lambda = enum_wrapper(source) || (source.respond_to?(:call) && source) || queue_wrapper(source)
+      @source = source.to_a unless @lambda # turn non-Enumerable-s into an Array
+      @runloop_queue = Thread::Queue.new if @lambda
       @mutex = mutex
       @index = -1
       @stopped = false
     end
 
-    def next
-      if producer?
+    def next(queue_for_thread = nil)
+      if @runloop_queue && queue_for_thread
+        return if @stopped
+        item = runloop_enq(queue_for_thread)
+        return if item == Stop
```

Comment:
why is it not like this ?
```suggestion
        item = runloop_enq(queue_for_thread)
        @stopped = (item == Stop)
        return if @stopped
```

Reply:
`@stopped` will be set in `JobFactory#runloop` .

This `JobFactory#next` may be called from some method at the same time.
So your `@stopped = (item == Stop)` needs exclusive control `Mutex#synchronize` .

Taking the value out of the `@lambda` and Setting the result to`@stopped` must be handled in the critical section or  be handled in one thread, I think.
Otherwise, the certain thread may clear `@stopped` flag. This could be a bug.

So previous implementation of Parallel uses `@mutex.synchronize` .
This PR's code handles `@lambda` and check `item == Stop` in `#runloop` by one thread, so `@mutex` is given, but it's not used.

Reply:
ah thx, yeah this is a tricky section :)
can you leave a bit of inline comment for the gotchas

Reply:
I have added comment to this code.